keps/0008-node-heartbeat.md

```diff
@@ -0,0 +1,374 @@
+      * [Split Node object into two parts at etcd level](#split-node-object-into-two-parts-at-etcd-level)
+      * [Delta compression in etcd](#delta-compression-in-etcd)
+      * [Replace etcd with other database](#replace-etcd-with-other-database)
+
+## Summary
+
+Node heartbeats are necessary for correct functioning of Kubernetes cluster.
+This proposal makes them significantly cheaper from both scalability and
+performance perspective.
+
+## Motivation
+
+While running different scalability tests we observed that in big enough clusters
+(more than 2000 noodes) with non-trivial number of images used by pods on all
+nodes (10-15), we were hitting etcd limits for its database size. That effectively
+means that etcd enters "alert mode" and stops accepting all write requests.
+
+The underlying root cause is combination of:
+
+- etcd keeping both current state and transaction log with copy-on-write
+- node heartbeats being pontetially very large objects (even though
+  the patch send over network is small, in etcd we store the whole Node
+  object) - they may easily exceed 15kB
+- Kubelet sending heartbeats every 10s
+
+This proposal presents a proper solution for that problem.
+
+
+Note that currently:
+
+- Lack of NodeStatus update for 40s results in NodeController marking node
+  as NotReady (pods are no longer scheduled on that node)
+- Lack of NodeStatus updates for 5m results in NodeController starting pod
+  evictions from that node
+
+We would like to preserve that behavior.
+
+
+### Goals
+
+- Reduce size of etcd by making node heartbeats cheaper
+
+### Non-Goals
+
+The following are nice-to-haves, but not primary goals:
+
+- Reduce resource usage (cpu/memory) of control plane (e.g. due to processing
+  less and/or smaller objects)
+- Reduce watch-related load on Node objects
+
+## Proposal
+
+We propose introducing a new `LeaderElection` CRD with the following schema:
+```
+type LeaderElection struct {
+  metav1.TypeMeta `json:",inline"`
+  // Standard object's metadata.
+  // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+  // +optional
+	ObjectMeta metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // Specification of the LeaderElection.
+  // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+  // +optional
+  Spec LeaderElectionSpec `json:"spec,omitempty"`
+}
+
+type LeaderElectionSpec struct {
+  HolderIdentity       string      `json:"holderIdentity"`
+  LeaseDurationSeconds int32       `json:"leaseDurationSeconds"`
+  AcquireTime          metav1.Time `json:"acquireTime"`
+  RenewTime            metav1.Time `json:"renewTime"`
+  LeaderTransitions    int32       `json:"leaderTransitions"`
+}
+```
+
+The Spec is effectively of already existing (and thus proved) [LeaderElectionRecord][].
+That would hopefully allow us go get directly to Beta.
+
+We will use CRD (as opposed to built-in API), because:
+
+- CRDs are `the` way to created new APIs
```

Comment:
I don't think this is quite accurate. It doesn't seem proper for a core kubernetes component to depend on the existence of a "custom" resource, which would be required to exist in all clusters to be functional

Reply:
I'm not the API-expert. I've been discussing it with @bgrant0607 and he explicitly said that it should be done with CRD.
Personally, I don't really have preference here, but I trust that Brian knows what he is saying.

Reply:
> I'm not the API-expert. I've been discussing it with @bgrant0607 and he explicitly said that it should be done with CRD. Personally, I don't really have preference here, but I trust that Brian knows what he is saying.

that doesn't seem proper... node heartbeating isn't an extension, add-on, or optional component... it's fundamental to a kubernetes cluster.

Reply:
@liggitt - is your concerns only about CRD vs built-in API or more?

To cite Brian on CRDs:
```
1. CRD is a mechanism for building APIs
2. Most new APIs will at least start as CRDs
3. Most new APIs will stay CRDs forever
```
But I will let @bgrant0607 to comment more on that.

Reply:
> is your concerns only about CRD vs built-in API or more?

That's my main concern. An API the kubelet depends on is an API the core repo has to define, document, build clients for, and support. Most of those are easier and more reliable for the kube-apiserver to do as a compiled-in type, rather than a CRD. Given that the schema stability and iteration time is tied to kube releases, I don't see many benefits to building these as a CRD, and see plenty of downsides.

Reply:
I actually see the arguments against CRDs as pretty weak.

Extension mechanisms need to be in the Nucleus so that we can use them to build the rest of the system. We're trying to make the apiserver dumb, generic functionality as much as possible. In particular, we're trying to make CRDs good enough to build APIs as good as built-in ones for most resources. And we'd like most new APIs to at least start as CRDs. 

This resource is just CRUD. It's implemented in annotations on ConfigMaps currently. We could just use that, with the ConfigMaps in a new namespace to reduce the watch cost.

If this were a built-in API, it would probably start as alpha, in which case Kubelet also couldn't depend on it. The node controller also can't depend on it, since we need to support version-skewed kubelets.

Reply:
If write path efficiency matters, CRDs are at least 2-5x more expensive than a proper api with generated types.

Prototyping as a CRD makes some sense.  Splitting other pieces out into their own CRD types makes sense.  But if heartbeating is significant enough to make special its unlikely CRD is enough

Reply:
Given that etcd has a special api specifically for leases, I’m having some trouble believing we should implement that same abstraction. it as a generic abstraction on top of the generic key value store.  I’d much prefer to leverage the underlying semantics of the store to make lease as efficient as possible.  Abstractions are great except when we push them too far.

Reply:
For instance, this is a great place to actually uses separate key ordered inline with nodes which is transactionally updated and synthsized into node or another resource as needed, preserving watch.

Reply:
@smarterclayton 
I did some first benchmarks of it and indeed CRDs vs built-in (which would mostly be json vs protobuf) would be ~5-6x slower. But, the object is fairly small, and even if that is 5x slower, it's still better the serialization of the whole Node object. Given that serialization/deserialization of Node objects is not our main painpoint in apiserver, I'm not that worried about that.

Re leases - I thought about it and I think that's actually not the best idea due to one reason:

Leases are not watchable (i.e. only creation and deletion of lease are watchable IIUC).
If we would use leases for that purpose, we would break all people who would like to watch for all heartbeats.
If we use regular API, they would just need to migrate to use this object instead.


> For instance, this is a great place to actually uses separate key ordered inline with nodes which is transactionally updated and synthsized into node or another resource as needed, preserving watch.

This is some variation of `Split Node object into two parts at etcd level` alternative mentioned below. Splitting object into two an synthetizing them to preserve a watch is pretty difficult. I fleshed some more thoughts about it in https://github.com/kubernetes/kubernetes/issues/14733#issuecomment-362217568 couple weeks ago. This would be a lot of very complicated code that wouldn't be very generic, I thinkg (@lavalamp for thoughts too).
You know that we already had bugs in cacher, and that is place that would have to be significantly changed for that and I guess we would have more.
So after more thinking about that, I'm quite against about presenting that in watch.

Reply:
One of the concerns I have for using CRD is: where & how these CRD would be maintained? e.g. Separate repo?

Reply:
The existing leader election code is in client-go. It could go there, or elsewhere in k/k.

Reply:
I initially wanted to have this somewhere in k/k. But it's not a strong preference - client-go sounds also fine for me.

Reply:
Clarified where to put that code in the doc.

Reply:
@wojtek-t Can you include your experimental benchmark on CRD vs. built-in into this KEP?  I am fine with CRD, at least we can start with CRD first. If the performance concern remains, we can graduate that to built-in later.

Reply:
Technically, so far I only did json vs protobuf benchmark on this particular object with some ~real data. I can create a better benchmark when I'm back from leave in couple days.

@dchen1107 that said - how exactly you want this to be incorporated into this KEP? Just the results? Or maybe link to PR to main repo? Or sth else?